### PR TITLE
Add --verify and --verify-only option to west flash

### PR DESCRIPTION
**Pull request by yashi**
These commits add `--verify` and `--verify-only` options to `west flash`.  These commits are only modifying `do_flash_elf()` for now.

Also I saw that `do_flash()` uses `self.load_cmd` and `self.verify_cmd` but `do_flash_elf()` doesn't.  I haven't look into it yet.

### CODE REVIEW

**Comment by mbolivar-nordic** [comment]
> Also I saw that `do_flash()` uses `self.load_cmd` and `self.verify_cmd` but `do_flash_elf()` doesn't. I haven't look into it yet.

That's because do_flash_elf() was added later and is kind of a hack.

I think it would be a bit better to make --use-elf play nicely with --cmd-verify if possible.

**Comment by abrodkin** [comment]
@yashi which Zephyr SDK version do you use for testing and what is your target platform?
I just tried it on ARC HSDK board with SDK versions 0.12.4 (latest stable) & 0.13.0 Alpha-1 (latest alpha) and in both cases `openocd` segfaults:
```
$ west -v -v -v flash --serial=25164200001f --use-elf --verify
args namespace: Namespace(command='flash', help=None, verbose=3, zephyr_base=None)
ZEPHYR_BASE=.../zephyr (origin: configfile)
config dir-fmt: build
build is a zephyr build directory
build dir: build
.../zephyr/build is a zephyr build directory
-- west flash: rebuilding
cmake version 3.16.3 is OK; minimum version is 3.13.1
Running CMake: /global/freeware/Linux/RHEL7/python-3.7.0/bin/cmake --build .../zephyr/build
ninja: no work to do.
-- west flash: using runner openocd
runner_config: RunnerConfig(build_dir='.../zephyr/build', board_dir='.../zephyr/boards/arc/hsdk', elf_file='.../zephyr/build/zephyr/zephyr.elf', hex_file=None, bin_file=None, gdb='.../zephyr-sdk/0.13.0-alpha-1/arc-zephyr-elf/bin/arc-zephyr-elf-gdb', openocd='.../zephyr-sdk/0.13.0-alpha-1/sysroots/x86_64-pokysdk-linux/usr/bin/openocd', openocd_search='.../zephyr-sdk/0.13.0-alpha-1/sysroots/x86_64-pokysdk-linux/usr/share/openocd/scripts')
runners.openocd: .../zephyr-sdk/0.13.0-alpha-1/sysroots/x86_64-pokysdk-linux/usr/bin/openocd -s .../zephyr/boards/arc/hsdk/support -s .../zephyr-sdk/0.13.0-alpha-1/sysroots/x86_64-pokysdk-linux/usr/share/openocd/scripts '-c set _ZEPHYR_BOARD_SERIAL 25164200001f' -f .../zephyr/boards/arc/hsdk/support/openocd.cfg -c init -c targets -c 'reset halt' -c 'load_image .../zephyr/build/zephyr/zephyr.elf' -c 'verify_image .../zephyr/build/zephyr/zephyr.elf' -c 'resume 0x0000000090001254' -c shutdown
Open On-Chip Debugger 0.10.0+dev-01508-gf79c90268-dirty (2021-05-05-17:15)
Licensed under GNU GPL v2
For bug reports, read
        http://openocd.org/doc/doxygen/bugs.html
25164200001f
Info : target has l2 cache enabled is enabled
Info : target has l2 cache enabled is enabled
Info : target has l2 cache enabled is enabled
Info : target has l2 cache enabled is enabled
Info : ftdi: if you experience problems at higher adapter clocks, try the command "ftdi_tdo_sample_edge falling"
Info : clock speed 10000 kHz
Info : JTAG tap: arc-em.cpu4 tap/device found: 0x200c24b1 (mfg: 0x258 (ARC International), part: 0x00c2, ver: 0x2)
Info : JTAG tap: arc-em.cpu3 tap/device found: 0x200824b1 (mfg: 0x258 (ARC International), part: 0x0082, ver: 0x2)
Info : JTAG tap: arc-em.cpu2 tap/device found: 0x200424b1 (mfg: 0x258 (ARC International), part: 0x0042, ver: 0x2)
Info : JTAG tap: arc-em.cpu1 tap/device found: 0x200024b1 (mfg: 0x258 (ARC International), part: 0x0002, ver: 0x2)
Info : starting gdb server for arc-em.cpu2 on 3333
Info : Listening on port 3333 for gdb connections
    TargetName         Type       Endian TapName            State
--  ------------------ ---------- ------ ------------------ ------------
 0  arc-em.cpu2        arcv2      little arc-em.cpu2        halted
 1  arc-em.cpu3        arcv2      little arc-em.cpu3        halted
 2  arc-em.cpu4        arcv2      little arc-em.cpu4        halted
 3* arc-em.cpu1        arcv2      little arc-em.cpu1        halted

Info : JTAG tap: arc-em.cpu4 tap/device found: 0x200c24b1 (mfg: 0x258 (ARC International), part: 0x00c2, ver: 0x2)
Info : JTAG tap: arc-em.cpu3 tap/device found: 0x200824b1 (mfg: 0x258 (ARC International), part: 0x0082, ver: 0x2)
Info : JTAG tap: arc-em.cpu2 tap/device found: 0x200424b1 (mfg: 0x258 (ARC International), part: 0x0042, ver: 0x2)
Info : JTAG tap: arc-em.cpu1 tap/device found: 0x200024b1 (mfg: 0x258 (ARC International), part: 0x0002, ver: 0x2)
L2 cache is present and not disabled
Invalidating L2 cache...
L2 cache has been flushed and invalidated.
19808 bytes written at address 0x90000000
564 bytes written at address 0x9000794c
downloaded 20372 bytes in 0.034362s (578.969 KiB/s)

FATAL ERROR: command exited with status -11: .../zephyr-sdk/0.13.0-alpha-1/sysroots/x86_64-pokysdk-linux/usr/bin/openocd -s .../zephyr/boards/arc/hsdk/support -s .../zephyr-sdk/0.13.0-alpha-1/sysroots/x86_64-pokysdk-linux/usr/share/openocd/scripts '-c set _ZEPHYR_BOARD_SERIAL 25164200001f' -f .../zephyr/boards/arc/hsdk/support/openocd.cfg -c init -c targets -c 'reset halt' -c 'load_image .../zephyr/build/zephyr/zephyr.elf' -c 'verify_image .../zephyr/build/zephyr/zephyr.elf' -c 'resume 0x0000000090001254' -c shutdown
=== Traceback (enabled by -vvv):
Traceback (most recent call last):
  File "/u/abrodkin/.local/lib/python3.7/site-packages/west/app/main.py", line 339, in run_command
    self.run_extension(args.command, argv)
  File "/u/abrodkin/.local/lib/python3.7/site-packages/west/app/main.py", line 409, in run_extension
    command.run(args, unknown, self.topdir, manifest=self.manifest)
  File "/u/abrodkin/.local/lib/python3.7/site-packages/west/commands.py", line 116, in run
    self.do_run(args, unknown)
  File ".../zephyr/scripts/west_commands/flash.py", line 29, in do_run
    do_run_common(self, my_args, runner_args)
  File ".../zephyr/scripts/west_commands/run_common.py", line 228, in do_run_common
    runner.run(command_name)
  File ".../zephyr/scripts/west_commands/runners/core.py", line 509, in run
    self.do_run(command, **kwargs)
  File ".../zephyr/scripts/west_commands/runners/openocd.py", line 132, in do_run
    self.do_flash_elf(**kwargs)
  File ".../zephyr/scripts/west_commands/runners/openocd.py", line 215, in do_flash_elf
    self.check_call(cmd)
  File ".../zephyr/scripts/west_commands/runners/core.py", line 606, in check_call
    subprocess.check_call(cmd, **kwargs)
  File "/global/freeware/Linux/RHEL7/python-3.7.0/lib/python3.7/subprocess.py", line 328, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['.../zephyr-sdk/0.13.0-alpha-1/sysroots/x86_64-pokysdk-linux/usr/bin/openocd', '-s', '.../zephyr/boards/arc/hsdk/support', '-s', '.../zephyr-sdk/0.13.0-alpha-1/sysroots/x86_64-pokysdk-linux/usr/share/openocd/scripts', '-c set _ZEPHYR_BOARD_SERIAL 25164200001f', '-f', '.../zephyr/boards/arc/hsdk/support/openocd.cfg', '-c', 'init', '-c', 'targets', '-c', 'reset halt', '-c', 'load_image .../zephyr/build/zephyr/zephyr.elf', '-c', 'verify_image .../zephyr/build/zephyr/zephyr.elf', '-c', 'resume 0x0000000090001254', '-c', 'shutdown']' died with <Signals.SIGSEGV: 11>.

```

If I run `openocd` separately under GDB I see this (I guess due to the fact that `openocd` in the SDK is stripped):
```
$ .../zephyr-sdk/0.13.0-alpha-1/sysroots/x86_64-pokysdk-linux/usr/bi
n/openocd -s .../zephyr/boards/arc/hsdk/support -s .../zephyr-sdk/0.13.0-alpha-1/sysroots/x86_64-pokys
dk-linux/usr/share/openocd/scripts '-c set _ZEPHYR_BOARD_SERIAL 25164200001f' -f .../zephyr/boards/arc/hsdk/support/openocd
.cfg -c init -c targets -c 'reset halt' -c 'load_image .../zephyr/build/zephyr/zephyr.elf' -c 'verify_image /SCRATCH/abrodk
in/Projects/sources/git/zephyrproject/zephyr/build/zephyr/zephyr.elf' -c 'resume 0x0000000090001254' -c shutdown

...

L2 cache has been flushed and invalidated.
19808 bytes written at address 0x90000000
564 bytes written at address 0x9000794c
downloaded 20372 bytes in 0.041367s (480.928 KiB/s)

Program received signal SIGSEGV, Segmentation fault.
0x0000000000000000 in ?? ()
(gdb) bt
#0  0x0000000000000000 in ?? ()
#1  0x000055555560276c in ?? ()
#2  0x00005555556029d6 in ?? ()
#3  0x0000555555629826 in ?? ()
#4  0x000055555577b987 in ?? ()
#5  0x0000000000000000 in ?? ()
```

I need to check if newer OpenOCD (from say upstream `master` or latest release) behaves any differently.
Though it shouldn't block this PR, it's quite unfortunate we cannot really validate it on one of those rare platforms which does use OpenOCD for loading Elf files on targets.

**Comment by yashi** [comment]
@abrodkin, I've never seen that segfault.  I've used both 0.12.x and 0.13 alpha as you did and also upstream version of it.  I did it against a customized version of Digilent Arty A7.

I've noticed that two commands use different path.  Do they reference the same image?
```
'load_image .../zephyr/build/zephyr/zephyr.elf'
'verify_image /SCRATCH/abrodkin/Projects/sources/git/zephyrproject/zephyr/build/zephyr/zephyr.elf'
```

Even if they don't, openocd should never segfaults.  If possible please build OpenOCD by yourself to take a backtrace.

**Comment by abrodkin** [comment]
@yashi thanks for the clarification. Paths are all OK as I tried to "obfuscate" them a bit an not spoil my folder structure, thus otherwise meaningless `...` ;)

I asked our in-house OpenOCD guru @EvgeniiDidin to look at what's going on there. So once he puts his hands on it we'll get more details here. If that one is not badly needed in the `main` branch, may we hold it and not merge until we may finally try it on our boards (after Evgeniy fixes segfault, hopefully sometime next week)?

**Comment by yashi** [comment]
Sure, no problem.  Thank you for testing! If you get a backtrace, let me know.  I might be any help.

**Comment by yashi** [comment]
@abrodkin, any progress?

**Comment by abrodkin** [comment]
@yashi sorry for that long delay. So thanks to @EvgeniiDidin we know the reason for these segfaults. Basically some call-backs are not implemented for ARC. In that case it's `.checksum_memory` (see https://github.com/zephyrproject-rtos/openocd/blob/zephyr-20210630/src/target/arc.c#L2504), called from https://github.com/zephyrproject-rtos/openocd/blob/zephyr-20210630/src/target/target.c#L2522:
```c
int target_checksum_memory(struct target *target, target_addr_t address, uint32_t size, uint32_t *crc)
{
...

	retval = target->type->checksum_memory(target, address, size, &checksum);
```

```
Program received signal SIGSEGV, Segmentation fault.
0x0000000000000000 in ?? ()
(gdb) bt
#0  0x0000000000000000 in ?? ()
#1  0x0000000000455e27 in target_checksum_memory (target=target@entry=0xa52450, address=2415919104, size=19952, crc=crc@entry=0x7fffffffcf64) at src/target/target.c:2522
#2  0x0000000000456066 in handle_verify_image_command_internal (cmd=0x7fffffffd040, verify=IMAGE_VERIFY) at src/target/target.c:3846
#3  0x000000000047e4ca in run_command (num_words=2, words=0xa80800, c=0x9f3ea0, context=0x9a7010) at src/helper/command.c:582
#4  exec_command (argv=0x7fffffffd140, argc=<optimized out>, c=0x9f3ea0, cmd_ctx=0x9a7010, interp=0x9a7070) at src/helper/command.c:950
#5  jim_command_dispatch (interp=0x9a7070, argc=<optimized out>, argv=0x7fffffffd140) at src/helper/command.c:1001
#6  0x00000000005b839a in JimInvokeCommand (interp=interp@entry=0x9a7070, objc=objc@entry=2, objv=objv@entry=0x7fffffffd140) at jim.c:10514
#7  0x00000000005b8fbc in Jim_EvalObj (interp=interp@entry=0x9a7070, scriptObjPtr=scriptObjPtr@entry=0x9e96d0) at jim.c:11006
#8  0x00000000005bb12b in Jim_EvalSource (interp=interp@entry=0x9a7070, filename=filename@entry=0x0, lineno=lineno@entry=0,
    script=script@entry=0x9e0db0 "verify_image build/zephyr/zephyr.elf") at jim.c:11298
#9  0x000000000047df5f in command_run_line (context=context@entry=0x9a7010, line=0x9e0db0 "verify_image build/zephyr/zephyr.elf") at src/helper/command.c:626
#10 0x000000000047f6b8 in parse_config_file (cmd_ctx=cmd_ctx@entry=0x9a7010) at src/helper/configuration.c:136
#11 0x00000000004053ae in openocd_thread (cmd_ctx=0x9a7010, argv=0x7fffffffd378, argc=18) at src/openocd.c:296
#12 openocd_main (argc=18, argv=0x7fffffffd378) at src/openocd.c:356
#13 0x00007ffff710d505 in __libc_start_main () from /lib64/libc.so.6
#14 0x0000000000404e9e in _start ()
```

So that was a purely technical part of pretty much no interest because of its simplicity.

Now here comes a bit more interesting part. Why that callback is needed in the end one may ask? What could be easier than reading back memory contents and calling it a day? The answer is - nothing. But it might be really painfully slow and take as long as it took to load the binary initially. And image of a few MiB's might easily take a handful of minutes to load. Have yet another coffee for verification step might be too much of caffeine at once. Is there a better way?

Surely there is. Instead of piping the binary over slow JTAG we may ask our X GHz target processor to do some CRC calculation in its DDR5 and then compare of the same checksum calculated on host. Looks nice!

But the problem is we need a tiny service application on target to do that. And quite probably such an applet was not a part of the image we need to verify. And those are arch-specific and could be found in `contrib/loaders/checksum`. But it's not only that. We also need to take care and act so as if there was nothing executed on the target CPU in between binary load and its subsequent execution. I.e. we need to save and then restore registers potentially polluted by the CRC calculation applet, IRQ mask etc. Moreover set a breakpoint or a watchpoint to catch the applet termination.

That said convenience here comes for a price of machine-specific OpenOCD updates, which are unfortunately missing for ARC as of today and I guess we won't get it implemented now...

Fortunately for poor folks like us there's a handy fallback to an olde good and simple full image read-back, see https://github.com/zephyrproject-rtos/openocd/blob/zephyr-20210630/src/target/target.c#L2529:
```c
int target_checksum_memory(struct target *target, target_addr_t address, uint32_t size, uint32_t *crc)
{
...
	retval = target->type->checksum_memory(target, address, size, &checksum);
	if (retval != ERROR_OK) {
...
		target_read_buffer(target, address, size, buffer);
...
		image_calculate_checksum(buffer, size, &checksum);
...
		free(buffer);
	}
```

And so with as simple change as:
```
diff --git a/src/target/arc.c b/src/target/arc.c
index 02eedc5d3..1c4e27fce 100644
--- a/src/target/arc.c
+++ b/src/target/arc.c
@@ -2475,6 +2475,13 @@ int arc_cache_flush(struct target *target)
        return ERROR_OK;
 }

+static int arc_checksum_memory(struct target *target,
+               target_addr_t address, uint32_t count,
+               uint32_t *checksum)
+{
+       return ERROR_FAIL;
+}
+
 /* ARC v2 target */
 struct target_type arcv2_target = {
        .name = "arcv2",
@@ -2501,7 +2508,7 @@ struct target_type arcv2_target = {

        .read_memory = arc_mem_read,
        .write_memory = arc_mem_write,
-       .checksum_memory = NULL,
+       .checksum_memory = arc_checksum_memory,
        .blank_check_memory = NULL,

        .add_breakpoint = arc_add_breakpoint,
**
```

we may do much better:
```
openocd -s .../git/openocd/tcl '-c set _ZEPHYR_BOARD_SERIAL 251642000147' -f boards/arc/hsdk/support/openocd.cfg -c init -c targets -c 'reset halt' -c 'load_image build/zephyr/zephyr.elf' -c 'verify_image build/zephyr/zephyr.elf' -c shutdown

L2 cache has been flushed and invalidated.
19952 bytes written at address 0x90000000
564 bytes written at address 0x900079dc
downloaded 20516 bytes in 0.030880s (648.807 KiB/s)

verified 20516 bytes in 0.114892s (174.383 KiB/s)

shutdown command invoked
```

BTW see how blazing fast it is still for real Zephyr image - 0.1s!

Let us do that change at least in Zephyr's OpenOCD fork and than that PR will be good to go!

**Comment by yashi** [comment]
Glad to hear from you, @abrodkin.
OpenOCD is [preparing for the next release](https://www.mail-archive.com/openocd-devel@lists.sourceforge.net/msg14292.html), v0.12.0, early next year.  Your change is local to an arch and trivial.  I'd say it'd be accepted before v0.12.0.   So, please post it to the upstream as well.  If you need help, please let me know.

**Comment by abrodkin** [comment]
@yashi feel free to try https://github.com/zephyrproject-rtos/openocd/pull/47 and comment on it.
@stephanosio @galak will need your attention on getting above PR merged and at some point SDK re-spinned.

**Review by mbolivar-nordic — APPROVED** [review]

**Comment by abrodkin** [comment]
@mbolivar-nordic @yashi what do you folks think of making that check enabled by default? That among other things will make it active when `twister` is being used. It's clear we may pass extra `west` option on `twister` invocation, but to me it looks interesting to force `twister` do that loaded binary integrity check as well.

**Comment by yashi** [comment]
@abrodkin I'm OK with it but that's another issue you have to ask someone else.  I don't think maintainers are happy to put more time one CI because it's already too long to finish.

**Review by abrodkin — APPROVED** [review]
Looks good to me now, though IMHO it might be more useful if verification is enabled by default.

**Comment by cfriedt** [comment]
@yashi - please rebase and fix conflicts.

**Comment by yashi** [comment]
I'll do it within a week.

**Comment by yashi** [comment]
@cfriedt Done, rebased.

**Pull request merged**
